### PR TITLE
Wayland clipboard

### LINKED ISSUE #3414: Wayland Clipboard support
The full version:

`wlfreerdp /u:<myusername> +clipboard  /v:192.168.0.59
[17:21:46:392] [6451:6451] [ERROR][com.winpr.library] - LoadLibraryA: /usr/lib/freerdp2/libcliprdr-client.so: cannot open shared object file: No such file or directory
`

This happens with the current version in HEAD on openSUSE Tumbleweed with Gnome 3.20 "on Wayland".

The file libcliprdr-client.so does not exist. 
The wlfreerdp was build on OBS with these options:
cmake \
    -DCMAKE_BUILD_TYPE=Release \
    -DCMAKE_INSTALL_PREFIX:PATH=%{_prefix} \
    -DCMAKE_INSTALL_LIBDIR:PATH=%{_lib} \
    -DCMAKE_SKIP_INSTALL_RPATH=ON \
    -DWITH_CUPS=ON \
    -DWITH_CLIENT=ON \
    -DWITH_PCSC=ON \
    -DWITH_ALSA=ON \
    -DWITH_PULSE=ON \
    -DWITH_DIRECTFB=OFF \
    -DWITH_FFMPEG=OFF \
    -DWITH_JPEG=ON \
    -DWITH_GSM=ON \
    -DWITH_GSTREAMER_1_0=ON \
    -DWITH_IPP=OFF \
    -DWITH_OPENSSL=ON \
    -DWITH_CHANNELS=ON \
    -DWITH_X11=ON \
    -DWITH_XCURSOR=ON \
    -DWITH_XEXT=ON \
    -DWITH_XKBFILE=ON \
    -DWITH_XI=ON \
    -DWITH_XINERAMA=ON \
    -DWITH_XRENDER=ON \
    -DWITH_XV=ON \
    -DWITH_ZLIB=ON \
    -DSTATIC_CHANNELS=OFF \
    -DWITH_SERVER=OFF \
    -DWITH_DEBUG_SCARD=OFF \
    -DWITH_SMARTCARD_INSPECT=ON \
    -DWITH_WAYLAND=ON \
    -DWITH_LIBSYSTEMD=ON \
    -DCMAKE_NO_BUILTIN_CHRPATH=ON \

### PR DESCRIPTION
**Pull request by akallabeth**
* Implements basic clipboard functionality (currently only text is working, even HTML does not)
* Fixes #3414

TODOS:
* Format conversions (to/from HTML, convert between different image formats)
* File copy (client to server, server to client requires fuse)

### CODE REVIEW

File: uwac/libuwac/uwac-clipboard.c [hunk 1]
```diff
@@ -0,0 +1,288 @@
...
+{
+	.target = data_source_target_handler,
+	.send = data_source_send_handler,
+	.cancelled = data_source_cancelled_handler
+};
+
+static UwacReturnCode UwacRegisterDeviceListener(UwacSeat* s)
+{
+	if (!s)
+		return UWAC_ERROR_INTERNAL;
```
Comment (hardening) [inline uwac/libuwac/uwac-clipboard.c]: uneeded check

File: uwac/libuwac/uwac-clipboard.c [hunk 2]
```diff
@@ -0,0 +1,288 @@
...
+	if (rc != UWAC_SUCCESS)
+		return rc;
+
+	event = (UwacClipboardEvent*)UwacDisplayNewEvent(s->display, UWAC_EVENT_CLIPBOARD_AVAILABLE);
+
+	if (!event)
+	{
+		assert(uwacErrorHandler(s->display, UWAC_ERROR_INTERNAL,
+		                        "failed to allocate a close event\n"));
+		return UWAC_ERROR_INTERNAL;
```
Comment (hardening) [inline uwac/libuwac/uwac-clipboard.c]: I guess the message is more `failed to allocate a clipboard available message`

File: uwac/libuwac/uwac-clipboard.c [hunk 3]
```diff
@@ -0,0 +1,288 @@
...
+#include <sys/epoll.h>
+
+/* paste */
+static void data_offer_offer(void* data, struct wl_data_offer* data_offer,
+                             const char* offered_mime_type)
+{
+	UwacSeat* seat = (UwacSeat*)data;
+
+	if (seat && !seat->ignore_announcement)
+	{
```
Comment (hardening) [inline uwac/libuwac/uwac-clipboard.c]: `seat` can't be NULL here except if there's a buggy piece of code elsewhere (not relying on user provided data).

File: uwac/libuwac/uwac-clipboard.c [hunk 4]
```diff
@@ -0,0 +1,288 @@
...
+		UwacClipboardEvent* event = (UwacClipboardEvent*)UwacDisplayNewEvent(seat->display,
+		                            UWAC_EVENT_CLIPBOARD_OFFER);
+
+		if (!event)
+		{
+			assert(uwacErrorHandler(seat->display, UWAC_ERROR_INTERNAL,
+			                        "failed to allocate a close event\n"));
```
Comment (hardening) [inline uwac/libuwac/uwac-clipboard.c]: copy and pasted message

File: client/Wayland/wlf_cliprdr.c [hunk 5]
```diff
@@ -0,0 +1,863 @@
...
+	response.msgFlags = (data) ? CB_RESPONSE_OK : CB_RESPONSE_FAIL;
+	response.dataLen = size;
+	response.requestedFormatData = data;
+	return clipboard->context->ClientFormatDataResponse(clipboard->context,
+	        &response);
+}
+
+BOOL wlf_cliprdr_handle_event(wfClipboard* clipboard, const UwacClipboardEvent* event)
+{
+	if (!clipboard || !event)
```
Comment (hardening) [inline client/Wayland/wlf_cliprdr.c]: Not sure these checks are needed.
Anyway the `if (!clipboard)` check is duplicated below.

File: client/Wayland/wlf_cliprdr.c [hunk 6]
```diff
@@ -0,0 +1,863 @@
...
+ * Function description
+ *
+ * @return 0 on success, otherwise a Win32 error code
+ */
+static UINT wlf_cliprdr_send_client_capabilities(wfClipboard* clipboard)
+{
+	CLIPRDR_CAPABILITIES capabilities;
+	CLIPRDR_GENERAL_CAPABILITY_SET generalCapabilitySet;
+	capabilities.cCapabilitiesSets = 1;
+	capabilities.capabilitySets = (CLIPRDR_CAPABILITY_SET*) &
```
Comment (hardening) [inline client/Wayland/wlf_cliprdr.c]: just nitpicking but it would be cool to adjust the styling here and few line below (line 319)

Comment (akallabeth) [reply client/Wayland/wlf_cliprdr.c]: I'd prefer #4350 finally being agreed upon and merged ;)

File: client/Wayland/wlf_cliprdr.c [hunk 7]
```diff
@@ -0,0 +1,863 @@
...
+		}
+	}
+
+	if (clipboard->responseMime != NULL)
+	{
+		clipboard->responseFile = fdopen(fd, "w");
+
+		if (clipboard->responseFile)
+			wlf_cliprdr_send_data_request(clipboard, clipboard->responseFormat);
+	}
```
Comment (hardening) [inline client/Wayland/wlf_cliprdr.c]: Shouldn't we print something if the `fdopen` fails ?

File: client/Wayland/wlf_cliprdr.c [hunk 8]
```diff
@@ -0,0 +1,863 @@
...
+	                                     formatList->numFormats, sizeof(CLIPRDR_FORMAT))))
+	{
+		WLog_Print(clipboard->log, WLOG_ERROR, "failed to allocate %"PRIuz" CLIPRDR_FORMAT structs",
+		           clipboard->numServerFormats);
+		return CHANNEL_RC_NO_MEMORY;
+	}
+
+	clipboard->numServerFormats = formatList->numFormats;
+
+	if (!clipboard->seat)
```
Comment (hardening) [inline client/Wayland/wlf_cliprdr.c]: This looks so unlikely that we should print a message too when we hit that case.

File: uwac/libuwac/uwac-clipboard.c (hunk 3 above)
Comment (akallabeth) [reply uwac/libuwac/uwac-clipboard.c]: True, will add an assert

**Pull request merged**